Calculate needed meteorological parameters

AlexDo1 opened the pull request:
I started to calculate some values that we need in the final footprint calculation.

The script is a little bit messy at the moment, I added some random comments just for myself (like literature and sources), I will clean this up later.

I had the idea to "outsource" some of the calculations into stand-alone functions which will be called in the footprint function but I don´t really know the perfect structure for that, maybe you could help me on that.

In this way we could use some of this functions (like relative humidity calculation, air density and others) in other bridget functions in the future, as most of them are basic, often needed meteorological parameters.
I guess tests and modifications would also be easier with that structure.

AlexDo1 committed:
Update footprint.py

mmaelicke commented:
> The script is a little bit messy at the moment, I added some random comments just for myself (like literature and sources), I will clean this up later.

It is also possible to put references into a numpy-style docstring. So I would suggest to keep them at least in comments. Then I can use them later...

> I had the idea to "outsource" some of the calculations into stand-alone functions which will be called in the footprint function but I don´t really know the perfect structure for that, maybe you could help me on that.

Yeah, good idea. Just place more functions into the same file. As long as they rely only on their passed arguments, we can easily put them into another file later. It is also wise to place the actual method implementation of Korman into a new function like:

```python
def footprint(z, v, w, ..., method='Korman'):
    if method.lower() == 'korman':
      return _footprint_korman(z, v, w, ...)
    elif method.lower() == 'other_guy':
      ...

# korman specific implementation 
def _footprint_korman(z, v, w ...):
    ...
```

Then we can easily implement other functions.

> In this way we could use some of this functions (like relative humidity calculation, air density and others) in other bridget functions in the future, as most of them are basic, often needed meteorological parameters.

Yeah, if we need them outside `footprint.py` we can place them into a top-level `util` or `meteo` submodule at any later point in time.

> I guess tests and modifications would also be easier with that structure.

Oh, yes! 

So short: split the code in as many functions in `footprint.py` as you like. I can restructure it at any later point in time.
I will give you a review later..

AlexDo1 committed:
Update footprint.py

AlexDo1 committed:
Update footprint.py

AlexDo1 commented:
I defined functions to calculate the needed meteorological values.

I also created a jupyter notebook to calculate the root mean square error between the given values in the TK2 data and the calculated values from the new functions. 
Should I move that file with the needed data into the bridget/tests folder?

mmaelicke commented:
> I also created a jupyter notebook to calculate the root mean square error between the given values in the TK2 data and the calculated values from the new functions.
> Should I move that file with the needed data into the bridget/tests folder?

Yes! Good idea. Maybe I can even use it to write a unittest from it. Code review will follow...

AlexDo1 committed:
Meteo test

Added the jupyter notebook and data to test the calculation of some meteorological values in the /tests folder

mmaelicke commented:
@AlexDo1 why is the sampleDataHeader.csv so large? Be aware, that we should not yet distribute the data samples. Please remove the File from the repo and send it by mail. Then I can include it locally.
In the unittest we will used randomly generated data and test against that. The result does not have to be meaningful, it should just always produce the same result. This way we can detect breaking changes.

mmaelicke commented on bridget/eddy/footprint.py:
```diff
@@ -1,3 +1,5 @@
+import numpy as np
+
```
In Python, place the import always after the first comment. The first comment is always the documentation for the whole file. e.g.:

```python
"""
Description of the whole file/module. 

- what is a footprint
- how to import
- how to use
"""
import numpy as np


def my_first_function():
    """
    documentation of the first function.
    """
    ...
```

You don't have to perfectly fill all the docstrings with content.

mmaelicke commented on bridget/eddy/footprint.py:
```diff
@@ -15,7 +17,242 @@
-def footprint(t_air, a, p, u, cov_uw, cov_vw, cov_wt, var_v, z = 2.0, method = "Kormann_Meixner"):
+
+
+"""
+Functions to calculate meteorological paramters which are needed in the footprint function.
+"""
```
put this in the module docstring at the first lines of the file. After that, no comments outside of functions anymore

mmaelicke commented on bridget/eddy/footprint.py:
```diff
@@ -45,4 +282,78 @@ def footprint(t_air, a, p, u, cov_uw, cov_vw, cov_wt, var_v, z = 2.0, method = "
+
+    #f.7) calculate the flux footprint [1/m2]
+    mu = (1 + m) / r
+
+
+
+
+
+
+    pass
```
I think a return statement is missing, right?

AlexDo1 replied:
Yes, I´m currently working on the footprint function, but I decided to do that in another branch.

AlexDo1 replied:
Regarding the constants: I think these are often empirical solutions for the calculation of the variables. 
But you are right, I also noticed that some numbers or even formulas are varying in the literature. It was often difficult to find a usable formula at all so I was happy to find something that at least worked.

Maybe we could create a new issue for that as it is some work to go through the literature again and check for that.

mmaelicke replied:
> Maybe we could create a new issue for that as it is some work to go through the literature again and check for that.

I don't think that the two of us need to decide which constant etc. to use. Let's just expose the contants as parameters with default valule, like you did with the k parameter. I we don't do that in this PR, we should open an issue to not forget about it

> Yes, I´m currently working on the footprint function, but I decided to do that in another branch.

 Perfect. I'll review your changes again.

mmaelicke reviewed (changes requested):
Great work! I really like your clean, well-documented programming style. Better than what I do usually :).
There are some minor comments about Python file structure. 
Also, please go through all hard-coded numbers and check if they can possibly be user-defined one day? Hardcoding g=9.81 is fine from my point of view, but are there no other possibilites for e.g. the numbers in L.182? If not fine, if yes place them as optional parameters into that specific function (no need to expose them in the footprint function itself). Although we still use the same numbers and the user has no chance to change them, we mark them as potentially changeable by placing them as optional parameters. Makes it easier in future.

AlexDo1 committed:
Delete sampleDataHeader.csv

AlexDo1 committed:
Update footprint.py

AlexDo1 committed:
Update footprint.py

mmaelicke reviewed (approved):
Nice work!
